src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs

```diff
@@ -2,12 +2,101 @@
 
 using System.Linq;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Test.Utilities;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests
 {
     public class IndexAndRangeTests : CompilingTestBase
     {
+        [Fact]
+        [WorkItem(31889, "https://github.com/dotnet/roslyn/issues/31889")]
+        public void ArrayRangeIllegalRef()
+        {
+            var comp = CreateCompilationWithIndexAndRange(@"
+public class C {
+    public ref int[] M(int[] arr) {
+        ref int[] x = ref arr[0..2];
+        M(in arr[0..2]);
+        M(arr[0..2]);
+        return ref arr[0..2];
+    }
+    void M(in int[] arr) { }
+}");
+            comp.VerifyDiagnostics(
+                // (4,27): error CS1510: A ref or out value must be an assignable variable
+                //         ref int[] x = ref arr[0..2];
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "arr[0..2]").WithLocation(4, 27),
+                // (5,14): error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
+                //         M(in arr[0..2]);
+                Diagnostic(ErrorCode.ERR_RefReturnLvalueExpected, "arr[0..2]").WithLocation(5, 14),
+                // (7,20): error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
+                //         return ref arr[0..2];
+                Diagnostic(ErrorCode.ERR_RefReturnLvalueExpected, "arr[0..2]").WithLocation(7, 20));
+        }
+
+        [Fact]
+        [WorkItem(31889, "https://github.com/dotnet/roslyn/issues/31889")]
+        public void ArrayRangeIllegalRefNoRange()
+        {
+            var comp = CreateCompilationWithIndex(@"
+public class C {
+    public void M(int[] arr) {
+        ref int[] x = ref arr[0..2];
```

Comment:
Shouldn't we have a test that verifies that we can, or cannot, take the ref of an array access when the indexer is of type `Index`?